winton_kafka_streams/processor/_stream_thread.py

```diff
@@ -60,25 +89,61 @@ def process(self):
 
         record = self.queue.get()
         self.context.currentRecord = record
+        self.current_timestamp = self.timestamp_extractor.extract(record, self.current_timestamp)
 
         # TODO: FIXME-  assumes only one topic (next two lines)
         self.context.currentNode = self.topology.sources[0]
         self.topology.sources[0].process(record.key(), record.value())
 
         self.consumedOffsets[(record.topic(), record.partition())] = record.offset()
+        self.commitOffsetNeeded = True
 
         self.context.currentRecord = None
+        self.context.currentNode = None
+
+        return True
+
+    def maybe_punctuate(self):
+        timestamp = self.current_timestamp
+
+        if timestamp is None:
+            return False
+
+        return self.punctuation_queue.may_punctuate(timestamp)
+
+    def punctuate(self, node, timestamp):
+        log.debug(f'Punctuating processor {node} at {timestamp}')
+        self.context.currentRecord = DummyRecord(timestamp)
+        self.context.currentNode = node
+        node.punctuate(timestamp)
+        self.context.currentRecord = None
+        self.context.currentNode = None
 
     def commit(self):
```

Comment:
This is the one important line. Make sure we call .flush() on the producer before we commit offsets, so we don't skip any messages.
The sad side effect is that this introduces a bit of lag (pretty noticeable), as it needs to flush the outwards buffer of messages. Which means we need to try a bit harder to not commit all the time.